Remove unnecessary code: was When generating constructor, trim known prefixes from argument name to get parameter name

chborl opened the pull request:
Fixes #33673 

When generating a constructor, the name of the constructor's parameter is based on the name of the argument that was passed to the constructor.  This PR removes any common prefixes  ("_", "m_", "s_", "t_") before creating the parameter name.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Utilities/ParameterName.cs:
```diff
@@ -51,7 +52,7 @@ public ParameterName(string nameBasedOnArgument, bool isFixed, NamingRule parame
                 // Otherwise, massage it a bit to be a more suitable match for
                 // how people actually writing parameters.
-                var trimmed = nameBasedOnArgument.TrimStart('_');
+                var trimmed = NamingStyle.StripCommonPrefixes(nameBasedOnArgument, out var _);
```
1. consider `out _`.
2. consider having an overload without the second parameter at all.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Utilities/ParameterName.cs:
```diff
@@ -2,6 +2,7 @@
 using System.Linq;
 using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
 using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.NamingStyles;
```
sort.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/NamingStyles/NamingStyle.cs:
```diff
@@ -346,7 +346,7 @@ private string CreateCompliantNameReusingPartialPrefixesAndSuffixes(string name)
         }
 
-        private static string StripCommonPrefixes(string name, out string prefix)
+        internal static string StripCommonPrefixes(string name, out string prefix)
```
can be public. generally we don't have internal members (except inside public types).

sharwell replied:
:memo: This isn't a rule for this project. Both options are available for internal types.

CyrusNajmabadi replied:
It's not a rule.  but it's the general pattern we follow for the majority of code in the codebase.

> f you would like this to be a rule, please file it as an analyzer request in dotnet/roslyn-analyzers and/or itemizing it in #31140 rather than adding it to reviews.

i think it would be better to flip things around.  If you'd prefer to not have code consistency brought up in reviews, then it would be good to make the rule and demonstrate that it's in effect.
'
That switches the burden around.  That way we keep the code consistent by default, and if we want to lighten teh PR process, people can put in the legwork.  If we go the way you mentioned, it will end up making it so that we have all these requests (which likely won't go anywhere) and in teh meantime code consistency slides down.

sharwell commented on src/Workspaces/Core/Portable/NamingStyles/NamingStyle.cs:
```diff
@@ -346,7 +346,13 @@ private string CreateCompliantNameReusingPartialPrefixesAndSuffixes(string name)
+        public static string StripCommonPrefixes(string name)
+        {
+            StripCommonPrefixes(name, out _);
+            return name;
```
❗️  This is just returning the argument without changes.

sharwell commented on src/Workspaces/Core/Portable/NamingStyles/NamingStyle.cs:
```diff
@@ -346,7 +346,13 @@ private string CreateCompliantNameReusingPartialPrefixesAndSuffixes(string name)
+        public static string StripCommonPrefixes(string name)
+        {
+            StripCommonPrefixes(name, out _);
+            return name;
```
❗️ The test _should_ have failed due to this mistake, but it did not. Prior to fixing this method, it would be good to update the test to a failing state so you can see that fixing this method fixes the test. (The changes to test and this method can be submitted as one commit; the separation is more to help you locally validate the change.)

chborl replied:
The tests were not failing because the desired changes were also being made in the `MakeCompliant` method called immediately after.  As a result, there are two lines of code before the `MakeCompliant` code that are no longer needed:
```csharp
 var trimmed = nameBasedOnArgument.TrimStart('_');
 BestNameForParameter = trimmed.Length > 0 ? trimmed.ToCamelCase() : nameBasedOnArgument;
```

I've updated this PR to remove those lines and added some tests to verify the behavior is still as expected.

chborl commented:
The tests were not failing because the desired changes were also being made in the `MakeCompliant` method called immediately after.  As a result, there are two lines of code before the `MakeCompliant` code that are no longer needed:
```csharp
 var trimmed = nameBasedOnArgument.TrimStart('_');
 BestNameForParameter = trimmed.Length > 0 ? trimmed.ToCamelCase() : nameBasedOnArgument;
```

I've updated this PR to remove those lines and added some tests to verify the behavior is still as expected.
